Sources/NIOPosix/NonBlockingFileIO.swift

```diff
@@ -712,3 +730,356 @@ public struct NIODirectoryEntry: Hashable {
+                return Int64(eof)
+            }
+        }
+    }
+
+    /// Write `buffer` to `fileHandle` in ``NonBlockingFileIO``'s private thread pool which is separate from Task pool.
+    ///
+    /// - parameters:
+    ///   - fileHandle: The `NIOFileHandle` to write to.
+    ///   - buffer: The `ByteBuffer` to write.
+    public func write(fileHandle: NIOFileHandle,
+                      buffer: ByteBuffer) async throws {
+        return try await self.write0(fileHandle: fileHandle, toOffset: nil, buffer: buffer)
+    }
+
+    /// Write `buffer` starting from `toOffset` to `fileHandle` in ``NonBlockingFileIO``'s private thread pool which is separate from Task pool.
+    ///
+    /// - parameters:
+    ///   - fileHandle: The `NIOFileHandle` to write to.
+    ///   - toOffset: The file offset to write to.
+    ///   - buffer: The `ByteBuffer` to write.
+    public func write(fileHandle: NIOFileHandle,
+                      toOffset: Int64,
+                      buffer: ByteBuffer) async throws {
+        return try await self.write0(fileHandle: fileHandle, toOffset: toOffset, buffer: buffer)
+    }
+
+    private func write0(fileHandle: NIOFileHandle,
+                        toOffset: Int64?,
+                        buffer: ByteBuffer) async throws {
+        let byteCount = buffer.readableBytes
+
+        guard byteCount > 0 else {
+            return
+        }
+
+        return try await self.threadPool.runIfActive {
+            try self.writeSync(fileHandle: fileHandle, byteCount: byteCount, toOffset: toOffset, buffer: buffer)
+        }
+    }
+
+    /// Open file at `path` on a private thread pool, run an operation given the file handle and region and then close the file handle.
+    ///
+    /// The open file operation runs on a private thread pool which is separate from Task pool.
+    ///    
+    /// - note: The reason this provides the `NIOFileHandle` and the `FileRegion` is that both the opening of a file as well as 
+    /// the querying of its size are blocking.
+    ///
+    /// - parameters:
+    ///     - path: The path of the file to be opened for reading.
+    ///     - body: operation to run with file handle and region
+    /// - returns: return value of operation
+    public func withOpenFile<Result>(path: String, _ body: (NIOFileHandle, FileRegion) async throws -> Result) async throws -> Result {
+        let file = try await self.threadPool.runIfActive {
+            let fh = try NIOFileHandle(path: path)
+            do {
+                let fr = try FileRegion(fileHandle: fh)
+                return UnsafeTransfer((handle: fh, region: fr))
+            } catch {
+                _ = try? fh.close()
+                throw error
+            }
+        }
+        do {
+            let result = try await body(file.wrappedValue.handle, file.wrappedValue.region)
+            try file.wrappedValue.handle.close()
+            return result
+        } catch {
+            try file.wrappedValue.handle.close()
```

Comment:
Do we want to throw the close error in this case or the bodies error? Also this might try to double close if the previous close threw.

Reply:
Yeah I'll fix the double throw. 

I don't know which is better to report the error from the body or the file close. The user might be using errors as a way to report state from the body and then will never get an error from the file close.

Reply:
Fixed double close. Left throwing error though